### PR TITLE
drivers: ieee802154: rf2xx: Move power table to devicetree and add Sub-Giga support

### LINKED ISSUE #47520: Support for sub-ghz channels in at86rf2xx radio driver
Currently the at86rf2xx radio driver does not support sub-ghz channels (e.g. 868 Mhz or 915Mhz). Microchip's [AT86RF215](https://www.microchip.com/en-us/product/AT86RF215) and [AT86RF212B](https://www.microchip.com/en-us/product/AT86RF212B) are capable of sub-ghz channels and would be a nice addition for long-range low power comms.

As it is the driver likely is not far off from supporting these chips - perhaps [RIOT's implementation](
https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/at86rf2xx.h) that supports all at86rf2xx devices could be used for reference.

@nandojve any advice on what would be required to support these other chips?

### PR DESCRIPTION
**Pull request by nandojve**
The current version of power table is hardcoded in the driver which is a problem when use devices in production. This change remove all hardcode from driver and re-implement the feature to allow people create a table which is defined in devicetree. The big advantage is that each board can define their own table based on lab tests and allows use of FEM devices inclusive.

In addition, add support to at86rf212[b] sub-giga devices. This work enables use of pages 0, 2 and 5 in accordance with IEEE-802.15.4/2003/2006/2011. The proprietary speeds can be object of future work.

CC: @markus-becker-tridonic-com @maxmclau

How to build and test with AVR-RZ600 modules on Atmel Boards

```
west build -b sam4e_xpro samples/net/sockets/echo_client -- -DSHIELD=atmel_rf2xx_legacy -DOVERLAY_CONFIG=overlay-802154-subg.conf
west build -b sam4s_xplained samples/net/sockets/echo_server -- -DSHIELD=atmel_rf2xx_xplained -DOVERLAY_CONFIG=overlay-802154-subg.conf
```

Note: Ethernet/MDIO drivers should be disabled at SAM4E-XPRO

### CODE REVIEW

File: dts/bindings/ieee802154/atmel,rf2xx.yaml [hunk 1]
```diff
@@ -38,3 +38,53 @@ properties:
...
+
+    tx-pwr-min:
+      type: uint8-array
+      default: [ 0x00, 0x00 ]
+      description: |
+        This value represent minimum normalized value in dBm for the transceiver.
+        This property must be used with tx-pwr-mapping. Note that this allows
+        use of external transceiver. The value is represented by two values where
+        first represent signal indication and second is the value itself. The
+        default value is set equal to tx-pwr-max to create a table with one entry.
```
Comment (galak) [inline dts/bindings/ieee802154/atmel,rf2xx.yaml:80]: which table has one entry in it?  The last sentence is confusing.

File: dts/bindings/ieee802154/atmel,rf2xx.yaml [hunk 2]
```diff
@@ -38,3 +38,53 @@ properties:
...
+
+    tx-pwr-max:
+      type: uint8-array
+      default: [ 0x00, 0x00 ]
+      description: |
+        This value represent maximum normalized value in dBm for the transceiver.
+        This property must be used with tx-pwr-mapping. Note that this allows
+        use of external transceiver. The value is represented by two values where
+        first represent signal indication and second is the value itself. The
+        default value is set equal to tx-pwr-min to create a table with one entry.
```
Comment (galak) [inline dts/bindings/ieee802154/atmel,rf2xx.yaml:90]: same comment as above.

File: dts/bindings/ieee802154/atmel,rf2xx.yaml [hunk 3]
```diff
@@ -38,3 +38,53 @@ properties:
...
         Specifies the MAC address that was assigned to the network
         device
+
+    channel-page:
+      type: int
+      enum:
+        - 0
+        - 2
+        - 5
+      default: 0
```
Comment (galak) [inline dts/bindings/ieee802154/atmel,rf2xx.yaml:48]: since you are using DT_*PROP_OR you should remove the defaults from the binding.

File: dts/bindings/ieee802154/atmel,rf2xx.yaml (hunk 1 above)
Comment (nandojve) [reply dts/bindings/ieee802154/atmel,rf2xx.yaml]: Add a more detailed description.

File: dts/bindings/ieee802154/atmel,rf2xx.yaml (hunk 2 above)
Comment (nandojve) [reply dts/bindings/ieee802154/atmel,rf2xx.yaml]: idem

File: dts/bindings/ieee802154/atmel,rf2xx.yaml (hunk 3 above)
Comment (nandojve) [reply dts/bindings/ieee802154/atmel,rf2xx.yaml]: Fixed

**Comment by nandojve** [comment]
Hi @galak ,

I was wondering if you have few time to revisit this.

**Review by rlubos — DISMISSED** [review]

File: dts/bindings/ieee802154/atmel,rf2xx.yaml [hunk 4]
```diff
@@ -38,3 +38,75 @@ properties:
...
+
+    tx-pwr-map:
+      type: uint8-array
+      default: [ 0x00 ]
+      description: |
+        This is the Transmission Power Mapping array used to comply with local
+        regulations. By default this value set an output power above 0dBm for all
+        transceivers. This property must be used with tx-pwr-min and tx-pwr-max
+        for normal operations. The number of elements in the array is defined by
+        CONFIG_IEEE802154_RF2XX_TX_PWR_MAP_ENTRIES Kconfig property. The max
```
Comment (mbolivar-nordic) [inline dts/bindings/ieee802154/atmel,rf2xx.yaml:67]: This introduces a dependency from Kconfig into DT, which is a bit strange, no? Why is this necessary?

Comment (nandojve) [reply dts/bindings/ieee802154/atmel,rf2xx.yaml:67]: My initial thought was to define this parameter as the basis for creating and validating the size of the table. As the drive allows the use of more than one radio, the table size is shared between them and it must be the same for both instances, as this parameter is in the device's configuration structure, I don't know if it's possible to get the size directly from the size array when there are two instances because it could have different values between the two devices.

**Review by carlescufi — DISMISSED** [review]

**Review by galak — DISMISSED** [review]
Agree with @mbolivar-nordic that having a dependency between dts and Kconfig as specified in the binding doesn't make sense.  We need to fix/clean that up.

File: dts/bindings/ieee802154/atmel,rf2xx.yaml (hunk 4 above)
Comment (mbolivar-nordic) [reply dts/bindings/ieee802154/atmel,rf2xx.yaml:67]: Can you do a BUILD_ASSERT that the table sizes are all the same within a DT_INST_FOREACH_STATUS_OKAY macro? Something like this:

```c
#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1
#define THE_TABLE_SIZE(inst) ARRAY_SIZE(....(inst))
#define CHECK_TABLE_SIZE(inst) BUILD_ASSERT(THE_TABLE_SIZE(inst) == THE_TABLE_SIZE(0));
DT_INST_FOREACH_STATUS_OKAY(CHECK_TABLE_SIZE)
#endif
```

**Comment by nandojve** [comment]
> Agree with @mbolivar-nordic that having a dependency between dts and Kconfig as specified in the binding doesn't make sense. We need to fix/clean that up.

Implemented check as suggested by @mbolivar-nordic and dropped Kconfig option.

**Review by mbolivar-nordic — DISMISSED** [review]
This review is for the DT binding only. Thanks for taking the time to make the change and for writing such great descriptions.

**Comment by nandojve** [comment]
@carlescufi , @galak , could you revisit this? I already got @mbolivar-nordic approval on devicetree.

**Comment by nandojve** [comment]
I was wondering if anyone could help with this PR.

**Comment by nandojve** [comment]
Fix access to table order and rebase.

**Review by rlubos — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_rf2xx.c [hunk 5]
```diff
@@ -66,47 +67,15 @@ static struct net_pkt rf2xx_ack_pkt = {
...
-	/* 0x2, */               /*  3.4 dBm */
-	/* 0x1, */               /*  3.7 dBm */
-	0x0                      /*  4   dBm: 4 */
-};
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) > 1
+#define THE_TABLE_SIZE(inst) DT_INST_PROP_LEN(inst, tx_pwr_table)
+#define CHECK_TABLE_SIZE(inst) \
+	BUILD_ASSERT(THE_TABLE_SIZE(inst) == TX_PWR_TABLE_SIZE, \
+		     "The TX PWR TABLE must be same size for all instances.");
+DT_INST_FOREACH_STATUS_OKAY(CHECK_TABLE_SIZE)
```
Comment (gmarull) [inline drivers/ieee802154/ieee802154_rf2xx.c:77]: note: you can allow multiple sizes by just defining the table outside of driver's config/data, and just take a pointer to the table + its size instead. See for example https://github.com/teslabs/spinner/blob/5591731a547a9d5ee50636360122e53625cacb4b/drivers/svpwm/svpwm_stm32.c#L298-L299

**Review by gmarull — COMMENTED** [review]
one minor nitpick

**Comment by nandojve** [comment]
Implemented suggestion from @gmarull to have different table sizes.

**Comment by nandojve** [comment]
Hi @mbolivar-nordic , @rlubos, @gmarull I was wondering if you could re approve PR.
Thank you!

**Review by rlubos — APPROVED** [review]

**Review by gmarull — APPROVED** [review]
